File: train.m
all_data = load("all_data.mat");


global classes = [0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 1 1 0 1 1 0 1 0 1 1 1 0 1 0 0];
global skin_supra = [4 5.3 4.3 5.6 4 3.8 4 3.2 4 4.5 4 3.5 5 5.6 5.2 4 5 4.8 4.2 7.3 5 7.3 4.2 4.8 2.2 5.6 6.83 5.8 4.67 3.67 5.25 3.67 5.83 6.67 4.17 5 5 4.6 8.8 6.83 7.166 7.66 4.667 7.3 6 7.33 8.16 7.6 7 14.3 4.3 11 4.17 6.67 8 4.17 2.83];
global skin_infra = [4.5 5.6 2.6 6.6 4.3 3.8 2.6 3.8 3.3 5 3 3 4.5 5 4.8 3.5 4 3.2 3 8 4.7 6.3 3.2 3.2 1 6 6.83 6.83 3.83 3.5 5.25 3.16 6.3 6.17 5.17 5.3 5.33 3.6 7 4.3 6 4 4.333 7.6 7 4.83 5.16 5.3 5 11.3 5.3 13 2.8 5 9 2.67 2.5];

function results = train_all(data, features, classifier, metrics)
  global classes
  global skin_supra
  global skin_infra
  
  num_patients = size(data.patients)(1);
  num_vowels = size(data.vowels)(2);
  num_metrics = size(metrics)(1);
  num_features = size(features)(1);
  
  % extract the features
  feature_data = cell(num_patients, num_vowels);
  for patient = 1:num_patients
    for vowel = 1:num_vowels
      num_samples = size(data.D{patient, vowel})(1);
      feature_data{patient, vowel} = zeros(num_samples, num_features);
      for sample = 1:num_samples
        for feature = 1:num_features
          feature_data{patient, vowel}(sample, feature) = features{feature}(data.D{patient, vowel}, skin_supra(1, patient), skin_infra(1, patient));
        end
      end
    end
  end
    
  results = cell(num_patients)(1);
  for left_out = 1:num_patients
    % combine data
    iter_data = [];
    test_data = [];
    for i = 1:num_patients
      for vowel = 1:num_vowels
        if i != left_out
          iter_data = [iter_data; feature_data{i, vowel}];
        else
          test_data = [test_data; feature_data{i, vowel}];
        end
      end
    end
    iter_classes = [ classes(1, 1:left_out-1), classes(1, left_out+1:end) ];
    
    % train the classifier
    test_predicted = classifier(iter_data, iter_classes', test_data);
    tp = 0;
    tn = 0;
    fp = 0;
    fn = 0;
    for i = 1:size(test_predicted)(1)
      if classes(1, left_out) == 1
        if test_predicted(i) == 1
          tp = tp + 1;
        else
          fn = fn + 1;
        end
      else
        if test_predicted(i) == 1
          fp = fp + 1;
        else
          tn = tn + 1;
        end
      end
    end
    
    % calculate the provided metrics
    results{left_out} = cell(num_metrics)(1);
    for metric = 1:size(metrics)(1)
      results{left_out}{metric} = metrics{metric}(tp, tn, fp, fn);
    end
    
  end
end

function labels = svm_classifier(X, Y, test)
  SVMModel = fitcsvm(X, Y, 'KernelFunction', 'rbf', 'Standardize', true);
  [labels, scores] = predict(SVMModel, test);
end

results = train_all(all_data, {@(channels, ss, si) mean(channels(1){1})}, @svm_classifier, {@(tp, tn, fp, fn) tp; @(tp, tn, fp, fn) tn; @(tp, tn, fp, fn) fp; @(tp, tn, fp, fn) fn});
disp(results)

File: train.py
import scipy.io
import numpy as np
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import precision_score, recall_score, accuracy_score
from collections import defaultdict
import h5py
from imblearn.over_sampling import SMOTE
import pickle

to_exclude = {
	# patient, vowel, sample
	2: {0: [47]},
	8: {0: [4, 5]},
	10: {2: 'all'},
	16: {0: [20, 27, 29, 31, 36]},
	28: {2: 'all'},
	36: {0: [52], 2: [21]},
	38: {1: [37]},
	50: {0: [37], 2: [13]},
	51: {1: [25]},
	54: 'all',
	55: 'all',
	56: 'all',
	57: 'all'
}

classes = np.array([0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0])

def separate_data(data_file, feature_data, to_exclude, left_out, num_patients, num_vowels, num_features):
	iter_data = np.array([]).reshape(0, num_features)
	test_data = np.array([]).reshape(0, num_features)
	iter_classes = np.array([]).reshape(0, 1)
	test_classes = np.array([]).reshape(0, 1)
	
	for i in range(num_patients):
		if i not in to_exclude or to_exclude[i] != 'all':
			for vowel in range(num_vowels):
				if i not in to_exclude or vowel not in to_exclude[i] or to_exclude[i][vowel] != 'all':
					if i in to_exclude and vowel in to_exclude[i]:
						to_remove = to_exclude[i][vowel]
					else:
						to_remove = []

					current_data = np.delete(np.transpose(data_file[feature_data[vowel, i]]), to_remove, axis=0)
					current_classes = np.delete(np.full((data_file[feature_data[vowel, i]].shape[1], 1), classes[i]), to_remove, axis=0)
					if not i == left_out:
						iter_data = np.concatenate((iter_data, current_data))
						iter_classes = np.concatenate((iter_classes, current_classes))
					else:
						test_data = np.concatenate((test_data, current_data))
						test_classes = np.concatenate((test_classes, current_classes))
						
	# remove infinite values (from dividing by 0)
	iter_data[~np.isfinite(iter_data)] = 0
	test_data[~np.isfinite(test_data)] = 0
	
	return iter_data, iter_classes, test_data, test_classes
	
	
def train_model(iter_data, iter_classes):
	X = iter_data
	y = iter_classes.ravel()

	# SMOTE
	rebalancing = SMOTE()
	X_res, y_res = rebalancing.fit_sample(X, y)
	
	# train classifier
	classifier = SVC()
	classifier.fit(X_res, y_res)
	
	return classifier
	

def train(file_name, variable_name):


	data_file = h5py.File(file_name, 'r')
	feature_data = data_file[variable_name]

	num_patients = feature_data.shape[1]
	num_vowels = feature_data.shape[0]
	num_features = data_file[feature_data[0, 0]].shape[0]

	# final metrics
	metrics = defaultdict(list)
	avg_metrics = defaultdict(int)

	# train and save full model
	iter_data, iter_classes, test_data, test_classes = separate_data(data_file, feature_data, to_exclude, -1, num_patients, num_vowels, num_features)
	classifier = train_model(iter_data, iter_classes)
	pickle.dump(classifier, open("model.pickle", 'wb'))

	num_trials = 0
	num_pos_trials = 0
	
	# leave one out cross-validation
	for left_out in range(num_patients):
		if not left_out in to_exclude or to_exclude[left_out] != 'all':
			print("PROGRESS", left_out)
			
			iter_data, iter_classes, test_data, test_classes = separate_data(data_file, feature_data, to_exclude, left_out, num_patients, num_vowels, num_features)
			classifier = train_model(iter_data, test_data)
			predicted_classes = classifier.predict(test_data)

			# evaluate metrics
			precision = precision_score(test_classes, predicted_classes)
			recall = recall_score(test_classes, predicted_classes)
			accuracy = accuracy_score(test_classes, predicted_classes)
			
			# store metrics
			metrics['precision'].append(precision)
			metrics['recall'].append(recall)
			metrics['accuracy'].append(accuracy)
			avg_metrics['precision'] += precision
			avg_metrics['recall'] += recall
			avg_metrics['accuracy'] += accuracy

			num_trials = num_trials + 1
			num_pos_trials = num_pos_trials + classes[left_out]

	avg_metrics['precision'] /= num_pos_trials
	avg_metrics['recall'] /= num_pos_trials
	avg_metrics['accuracy'] /= num_trials

	return (metrics, avg_metrics)

#metrics, avg_metrics = train('all_features', 'feature_data')
#metrics, avg_metrics = train('featuresV1.mat', 'feature_data')
#metrics, avg_metrics = train('featuresV2.mat', 'feature_data')
metrics, avg_metrics = train('featuresV4.mat', 'feature_data')
print(metrics)
print(avg_metrics)
